indico/modules/events/management/controllers/protection.py

```diff
@@ -71,34 +101,85 @@ class RHEventProtection(RHManageEventBase):
-                     'own_no_access_contact': self.event.own_no_access_contact}, **coordinator_privs)
+                     'own_no_access_contact': self.event.own_no_access_contact, 'permissions': permissions},
+                    **coordinator_privs)
 
     def _update_session_coordinator_privs(self, form):
         data = {field: getattr(form, field).data for field in form.priv_fields}
         update_session_coordinator_privs(self.event, data)
+
+    def _get_principal_permissions(self, principal):
+        """Retrieve a set containing the valid permissions of a principal."""
+        permissions = set()
+        if principal.full_access:
+            permissions.add(FULL_ACCESS_PERMISSION)
+        if principal.read_access:
+            permissions.add(READ_ACCESS_PERMISSION)
+        available_permissions = get_permissions_info()[0]
+        return permissions | (set(principal.permissions) & set(available_permissions))
+
+    def _update_permissions(self, current, new):
+        """Handle the updates of permissions and creations/deletions of acl principals.
+
+        :param current: A dict mapping principals and a set with its current permissions
+        :param new: A dict mapping principals and a set with its new permissions
+        """
+        def _get_splitted_permissions(_permissions):
+            full_access_permission = FULL_ACCESS_PERMISSION in _permissions
+            if full_access_permission:
+                _permissions.remove(FULL_ACCESS_PERMISSION)
+            read_access_permission = READ_ACCESS_PERMISSION in _permissions
+            if read_access_permission:
+                _permissions.remove(READ_ACCESS_PERMISSION)
+            return full_access_permission, read_access_permission, _permissions
+
+        for principal, permissions in current.viewitems():
+            if principal not in new:
+                self.event.update_principal(principal, False, False, del_permissions=current[principal])
+            elif permissions != new[principal]:
+                full_access, read_access, new_permissions = _get_splitted_permissions(new[principal])
+                current_permissions = current[principal] - {FULL_ACCESS_PERMISSION, READ_ACCESS_PERMISSION}
+                add_permissions = new_permissions - current_permissions
+                del_permissions = current_permissions - new_permissions
+                self.event.update_principal(principal, read_access, full_access, add_permissions=add_permissions,
+                                            del_permissions=del_permissions)
+        new_principals = set(new) - set(current)
+        for p in new_principals:
+            full_access, read_access, permissions = _get_splitted_permissions(new[p])
+            self.event.update_principal(p, read_access, full_access, add_permissions=permissions)
+
+
+class RHEventPermissionsDialog(RHManageEventBase):
+    def _process(self):
+        principal = json.loads(request.args['principal'])
```

Comment:
JSON in the query string is a bit ugly ;) I'd consider using a POST request... You just need to add `data-method=POST` and the dialog will be loaded with POST instead of GET.